src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs

```diff
@@ -14,105 +14,97 @@ namespace System.Buffers
 
-                data = startSegment.Memory;
+                // Bounds check
+                #if DEBUG
+                Debug.Assert(positionSegment == _sequenceEnd.GetObject() || _sequenceStart.GetObject() != _sequenceEnd.GetObject());
 
-                if (startSegment != endObject)
+                ReadOnlySequenceSegment<T> startSegment = Unsafe.As<ReadOnlySequenceSegment<T>>(_sequenceStart.GetObject());
+                ReadOnlySequenceSegment<T> endSegment = Unsafe.As<ReadOnlySequenceSegment<T>>(_sequenceEnd.GetObject());
+
+                // startSegment.RunningIndex + startIndex <= positionSegment.RunningIndex + positionIndex &&
+                // positionSegment.RunningIndex + positionIndex <= endSegment.RunningIndex + endIndex
+                // Rearranged to avoid overflow
+                Debug.Assert(startSegment.RunningIndex - positionSegment.RunningIndex <= GetIndex(position) - GetIndex(_sequenceStart) ||
+                             GetIndex(position) - GetIndex(_sequenceEnd) <= endSegment.RunningIndex - positionSegment.RunningIndex);
+                #endif
+
+                memory = positionSegment.Memory;
+                if (positionSegment != _sequenceEnd.GetObject())
                 {
-                    ReadOnlySequenceSegment<T> nextSegment = startSegment.Next;
 
-                    if (nextSegment == null && startSegment != endObject)
-                        ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
+                    ReadOnlySequenceSegment<T> nextSegment = positionSegment.Next;
+                    Debug.Assert(nextSegment != null);
 
                     next = new SequencePosition(nextSegment, 0);
-                    length = data.Length - startIndex;
+                    endIndex = memory.Length;
+                    //length = memory.Length - positionIndex;
                 }
             }
-            else if (type == SequenceType.Array)
+            else
             {
-                Debug.Assert(startObject is T[]);
+                // Bounds check
+                Debug.Assert(positionObject == _sequenceEnd.GetObject());
+                Debug.Assert(GetIndex(_sequenceStart) <= GetIndex(position) && GetIndex(position) <= GetIndex(_sequenceEnd));
 
-                data = new ReadOnlyMemory<T>(Unsafe.As<T[]>(startObject));
-            }
-            else if (type == SequenceType.MemoryManager)
-            {
-                Debug.Assert(startObject is MemoryManager<T>);
+                if (type == SequenceType.Array)
+                {
+                    Debug.Assert(positionObject is T[]);
 
-                data = (Unsafe.As<MemoryManager<T>>(startObject)).Memory;
-            }
-            else if (typeof(T) == typeof(char) && type == SequenceType.String)
-            {
-                Debug.Assert(startObject is string);
+                    memory = new ReadOnlyMemory<T>(Unsafe.As<T[]>(positionObject));
```

Comment:
We don't need Unsafe code here and we can afford to take the performance hit of casting it (to get type safety).

If we re-write the TryGetBuffer method as follows, we get type safety, and we are still faster than the original:
```C#
[MethodImpl(MethodImplOptions.AggressiveInlining)]
private bool TryGetBuffer(in SequencePosition position, out ReadOnlyMemory<T> memory, out SequencePosition next)
{
    object positionObject = position.GetObject();
    next = default;

    if (positionObject == null)
    {
        memory = default;
        return false;
    }

    SequenceType type = GetSequenceType(position);
    object endObject = _sequenceEnd.GetObject();
    int startIndex = GetIndex(position);
    int endIndex = GetIndex(_sequenceEnd);

    if (type == SequenceType.MultiSegment)
    {
        Debug.Assert(positionObject is ReadOnlySequenceSegment<T>);
                
        //ReadOnlySequenceSegment<T> startSegment = Unsafe.As<ReadOnlySequenceSegment<T>>(positionObject);
        ReadOnlySequenceSegment<T> startSegment = (ReadOnlySequenceSegment<T>)positionObject;

        if (startSegment != endObject)
        {
            ReadOnlySequenceSegment<T> nextSegment = startSegment.Next;

            if (nextSegment == null)
                ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();

            next = new SequencePosition(nextSegment, 0);
            memory = startSegment.Memory.Slice(startIndex);
        }
        else
        {
            memory = startSegment.Memory.Slice(startIndex, endIndex - startIndex);
        }
    }
    else if (type == SequenceType.Array)
    {
        if (positionObject != endObject)
            ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();

        Debug.Assert(positionObject is T[]);

        //memory = new ReadOnlyMemory<T>(Unsafe.As<T[]>(positionObject), startIndex, endIndex - startIndex);
        memory = new ReadOnlyMemory<T>((T[])positionObject, startIndex, endIndex - startIndex);
    }
    else if (typeof(T) == typeof(char) && type == SequenceType.String)
    {
        if (positionObject != endObject)
            ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();

        Debug.Assert(positionObject is string);

        //memory = (ReadOnlyMemory<T>)(object)(Unsafe.As<string>(positionObject)).AsMemory(startIndex, endIndex - startIndex);
        memory = (ReadOnlyMemory<T>)(object)((string)positionObject).AsMemory(startIndex, endIndex - startIndex);
    }
    else // type == SequenceType.MemoryManager
    {
        if (positionObject != endObject)
            ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();

        Debug.Assert(type == SequenceType.MemoryManager);
        Debug.Assert(positionObject is MemoryManager<T>);

        //memory = Unsafe.As<MemoryManager<T>>(positionObject).Memory.Slice(startIndex, endIndex - startIndex);
        memory = ((MemoryManager<T>)positionObject).Memory.Slice(startIndex, endIndex - startIndex);
    }

    return true;
}
```
![image](https://user-images.githubusercontent.com/6527137/38413879-ccd15f98-3941-11e8-9c0e-7bf82aab02bb.png)


cc @GrabYourPitchforks, @jkotas, @davidfowl

Reply:
If we check that positionObject == endObject that means that unsafe code will be correct always.

Reply:
In single segment we can just ignore positionObject and get memory directly from endObject with unsafe code.

Reply:
> If we check that positionObject == endObject means that unsafe code will be correct always.

That is true. Good point. 

> In single segment we can just ignore positionObject and get memory directly from endObject with unsafe code.

I am not completely convinced of this. Some tests will boost my confidence in that.

Reply:
FYI, @stephentoub, brought up an interesting issue the unsafe type casting can cause due to struct tearing: https://github.com/dotnet/corefx/issues/28920